lib/logstash/inputs/snmptrap.rb

```diff
@@ -47,26 +47,44 @@ def register
     end
   end # def register
 
-  public
   def run(output_queue)
     begin
       # snmp trap server
       snmptrap_listener(output_queue)
     rescue => e
       @logger.warn("SNMP Trap listener died", :exception => e, :backtrace => e.backtrace)
-      sleep(5)
-      retry
+      stoppable_call(20) { sleep(0.25) }
```

Comment:
what we/I did in other plugins is to actually hold an instance of the current thread in the run method and use it to wakeup via an atomic value.  I guess both does the same at the end of the day.